File: other/Solution42.java
import java.util.Stack;

class Solution42 {
    public int trap(int[] height) {
    	int sum = 0;
    	Stack<int[]> stack = new Stack<>();
    	for(int i = 0; i < height.length; i++){
    		if(stack.empty() || stack.peek()[0] > height[i]){
    			stack.push(new int[]{height[i], i});
    			continue;
    		}
    		int last = stack.pop()[0];
    		while(!stack.empty() && stack.peek()[0] <= height[i]){
    			int[] peek = stack.peek();
    			int distance = i - peek[1] - 1;
    			sum += (Math.min(peek[0], height[i]) - last) * distance;
    			last = stack.pop()[0];
    		}
    		if(!stack.empty()){
    			int distance = i - stack.peek()[1] - 1;
    			sum += (Math.min(height[i], stack.peek()[0]) - last) * distance;
			}
    		stack.push(new int[]{height[i], i});
    	}
    	return sum;
    }

    public static void main(String[] args){
        Solution42 s = new Solution42();
        int[] height = {4,2,0,3,2,5};
        System.out.println(s.trap(height));
    }
}

File: C++/Solution42.cpp
#include<iostream>
#include<vector>

#include<stack>
using namespace std;

class Solution42 {
public:
    int trap(vector<int>& height) {
        stack<int> stk;
        int sum = 0;
        for(int i = 0; i < height.size(); i++){
            if(stk.empty() || height[stk.top()] > height[i]){
                stk.push(i);
                continue;
            }
            int last = 0;
            while(!stk.empty() && height[stk.top()] <= height[i]){
                int index = stk.top();
                int h = height[index];
                sum += (h - last) * (i - index - 1);
                last = h;
                stk.pop();
            }
            if(!stk.empty()){
                int index = stk.top();
                int h = height[index];
                sum += (height[i] - last) * (i - index - 1);
            }
            stk.push(i);
        }
        return sum;
    }
};

int main(){
    Solution42 s;
    vector<int> test = {0,1,0,2,1,0,1,3,2,1,2,1};
    cout << s.trap(test);
    return 0;
}